File: backend/scripts/report_class.py
from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle
from reportlab.lib.units import cm
from reportlab.pdfbase import pdfmetrics

from reportlab.pdfbase.ttfonts import TTFont
from reportlab.pdfgen import canvas
from reportlab.lib.enums import TA_LEFT, TA_CENTER

from reportlab.pdfbase import pdfmetrics

print(111111111111111111, pdfmetrics.getRegisteredFontNames())


import os
import sys
import json
import io

font_dir = os.path.dirname(os.path.abspath(__file__))

pdfmetrics.registerFont(TTFont('Helvetica', os.path.join(font_dir, 'ofont.ru_DejaVu Sans.ttf')))
pdfmetrics.registerFont(TTFont('Helvetica-Bold', os.path.join(font_dir, 'ofont.ru_DejaVu Sans.ttf')))

def convert_to_five_point(percent):
    if percent < 40: return 2
    elif 40 <= percent < 60: return 3
    elif 60 <= percent < 80: return 4
    else: return 5

def generate_test_report(class_name: str, date: str, test_name: str, students_data: list, grading_system: list, output_filename: str):
    doc = SimpleDocTemplate(
        output_filename,
        pagesize=A4,
        topMargin=1.5 * cm,
        leftMargin=2 * cm,
        rightMargin=2 * cm,
        bottomMargin=4 * cm,
    )
    story = []

    # Стили для текста
    styles = getSampleStyleSheet()
    title_style = ParagraphStyle(
        "TitleStyle",
        parent=styles["Title"],
        fontName="Helvetica-Bold",
        fontSize=16,
        alignment=1,
    )
    subtitle_style = ParagraphStyle(
        "SubtitleStyle",
        parent=styles["Heading2"],
        fontName="Helvetica",
        fontSize=14,
        alignment=1,
    )
    body_style = ParagraphStyle(
        "BodyStyle",
        parent=styles["BodyText"],
        fontName="Helvetica",
        fontSize=12,
    )

    table_header_style = ParagraphStyle(
        "TableHeader",
        parent=styles["Normal"],
        fontName="Helvetica-Bold",
        fontSize=12,
        alignment=TA_CENTER
    )

    # Стили для таблицы
    table_text_style = ParagraphStyle(
        "TableText",
        parent=styles["BodyText"],
        fontName="Helvetica",
        fontSize=12,
        leading=12,
        alignment=TA_LEFT,
        wordWrap='LTR'
    )

    class FooterCanvas(canvas.Canvas):
        def __init__(self, *args, **kwargs):
            canvas.Canvas.__init__(self, *args, **kwargs)
            self.pages = []

        def showPage(self):
            self.pages.append(dict(self.__dict__))
            self._startPage()

        def save(self):
            for page in self.pages:
                self.__dict__.update(page)
                self.draw_canvas()
                canvas.Canvas.showPage(self)
            canvas.Canvas.save(self)

        def draw_canvas(self):
            self.saveState()
            self.setFont('Helvetica', 10)
            self.drawRightString(A4[0] - 2 * cm, 2 * cm, f"Дата: {date}")
            self.restoreState()

    # Шапка
    story.append(Paragraph(f"Отчёт по успеваемости класса <b>{class_name}</b>", title_style))
    story.append(Spacer(1, 0.2 * cm))
    story.append(Paragraph(f"Тест «{test_name}»", title_style))
    story.append(Spacer(1, 0.2 * cm))

    # Разделительная линия
    story.append(Spacer(1, 0.2 * cm))
    line = Table([[""]], colWidths=[17 * cm], rowHeights=[1])
    line.setStyle(TableStyle([("LINEABOVE", (0, 0), (-1, -1), 1, colors.black)]))
    story.append(line)
    story.append(Spacer(1, 0.5 * cm))

    # Вычисляем score для каждого ученика
    for student in students_data:
        if "score" not in student:
            correct_answers = sum(student["answers"])
            total_questions = len(student["answers"])
            student["score"] = round((correct_answers / total_questions) * 100) if total_questions > 0 else 0

    # Средняя оценка по классу
    avg_score = sum(student["score"] for student in students_data) / len(students_data)
    story.append(Paragraph(f"Средняя оценка по классу — {avg_score:.1f}%", body_style))
    story.append(Spacer(1, 1 * cm))

    # Сортируем учеников по aruco_num (как числам, а не строкам)
    sorted_students = sorted(students_data, key=lambda x: int(x["aruco_num"]))

    # Подготовка данных для таблицы
    table_headers = [
        Paragraph("№", table_header_style),
        Paragraph("Ученики", table_header_style),
        Paragraph("Ответы", table_header_style),
    ]

    # Добавляем колонки в зависимости от выбранной системы оценивания
    if 'percent' in grading_system:
        table_headers.append(Paragraph("Оценка (%)", table_header_style))
    if 'five-point' in grading_system:
        table_headers.append(Paragraph("Оценка (5-балльная)", table_header_style))

    table_data = [table_headers]

    for student in sorted_students:
        answer_symbols = "".join(["+" if ans else "-" for ans in student["answers"]])
        row = [
            Paragraph(student["aruco_num"], table_text_style),
            Paragraph(student["name"], table_text_style),
            Paragraph(answer_symbols, table_text_style),
        ]

        if 'percent' in grading_system:
            row.append(Paragraph(f"{student['score']}", table_header_style))
        if 'five-point' in grading_system:
            five_point_score = convert_to_five_point(student['score'])
            row.append(Paragraph(str(five_point_score), table_header_style))

        table_data.append(row)

    table = Table(table_data, colWidths=[1 * cm, 5.3 * cm, 6 * cm, 2.8 * cm, 3.4 * cm])
    table.setStyle(TableStyle([
        ("FONTNAME", (0, 0), (-1, -1), "Helvetica"),
        ("FONTSIZE", (0, 0), (-1, -1), 12),
        ("BACKGROUND", (0, 0), (-1, 0), colors.lightgrey),
        ("TEXTCOLOR", (0, 0), (-1, 0), colors.black),
        ("ALIGN", (0, 0), (-1, -1), "LEFT"),  # Все колонки по центру
        ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
        ("GRID", (0, 0), (-1, -1), 0.7, colors.black),
        ("BOLD", (0, 0), (-1, 0), True),
    ]))
    story.append(table)

    doc.multiBuild(story, canvasmaker=FooterCanvas)


def main():
    data = json.load(io.TextIOWrapper(sys.stdin.buffer, encoding='utf-8'))

    generate_test_report(
        class_name=data['class_name'],
        date=data['date'],
        test_name=data['test_name'],
        students_data=data['students_data'],
        grading_system=data['grading_system'],
        output_filename=data['output_filename']
    )
    print(data['class_name'])

if __name__ == "__main__":
    main()

# # Отчёт по классу по выбранному тесту
# if __name__ == "__main__":
#     class_name = "10-А"
#     # Дата проведения теста, можно брать дату проведения последнего вопроса из таблицы taken_question_answers
#     date = "11/04/2025"
#     test_name = "Алгебра: Квадратные уравнения"
#     grading_system = ['percent', 'five-point']
#     students_data = [
#         {"aruco_num": "2", "name": "Хохлов Тимофей", "answers": [False, True, True, False, True]},
#         {"aruco_num": "1", "name": "Иванов Иван", "answers": [True, True, False, True, False]},
#         {"aruco_num": "3", "name": "Петров Пётр", "answers": [True, True, True, True, True]},
#         {"aruco_num": "5", "name": "Сидорова Анна", "answers": [False, True, False, True, False]},
#         {"aruco_num": "4", "name": "Кузнецов Алексей", "answers": [True, False, True, False, True]},
#     ]

#     generate_test_report(class_name, date, test_name, students_data, grading_system, "test_report.pdf")
#     print("Отчёт успешно сгенерирован!")

File: backend/scripts/report_class.js
const PDFDocument = require('pdfkit');
// const fs = require('fs');
const path = require('path');

// Регистрация шрифтов (если нужно, укажи свои пути)
const fontDir = __dirname;
const fontPaths = {
    Helvetica: path.join(fontDir, 'Helvetica.ttf'),
    'Helvetica-Bold': path.join(fontDir, 'Helvetica-Bold.ttf')
};

// Конвертация процентов в 5-балльную систему
function convertToFivePoint(percent) {
    if (percent < 40) return 2;
    else if (percent < 60) return 3;
    else if (percent < 80) return 4;
    else return 5;
}

// Основная функция генерации отчёта
function generateTestReport({ class_name, date, test_name, students_data, grading_system }) {
    const doc = new PDFDocument({
        size: 'A4',
        margins: {
            top: 1.5 * 28.35,
            left: 2 * 28.35,
            right: 2 * 28.35,
            bottom: 2 * 28.35
        }
    });

    const chunks = [];
    doc.on('data', (chunk) => chunks.push(chunk));

    // Переменная для отслеживания текущей позиции Y
    let currentY = doc.y;

    // Функция для проверки и добавления новой страницы
    const checkPageBreak = (requiredHeight) => {
        const availableHeight = doc.page.height - doc.page.margins.bottom - currentY;
        if (requiredHeight > availableHeight) {
            addFooter(); // Добавляем футер перед новой страницей
            doc.addPage();
            currentY = doc.page.margins.top;
            return true;
        }
        return false;
    };

    // Функция добавления футера
    const addFooter = () => {
        const footerText = `Дата формирования отчёта: ${date}`;
        const footerFontSize = 10;

        doc.save()
            .font(fontPaths.Helvetica)
            .fontSize(footerFontSize)
            .text(footerText, doc.page.width - doc.page.margins.right - doc.widthOfString(footerText), doc.page.height - doc.page.margins.bottom - 15, { align: 'right' })
            .restore();
    };

    // Заголовок
    doc.font(fontPaths['Helvetica-Bold']).fontSize(16).text(`Отчёт по успеваемости класса ${class_name}`, {
        align: 'center'
    });
    doc.moveDown(0.2);
    doc.font(fontPaths['Helvetica-Bold']).fontSize(16).text(`Тест «${test_name}»`, {
        align: 'center'
    });
    doc.moveDown(0.2);
    currentY = doc.y;

    // Разделительная линия
    doc.moveTo(2 * 28.35, currentY)
        .lineTo(2 * 28.35 + 17 * 28.35, currentY)
        .stroke();
    doc.moveDown(0.5);
    currentY = doc.y;

    // Вычисляем score для каждого ученика
    students_data.forEach((student) => {
        if (!student.score) {
            const correctAnswers = student.answers.filter(Boolean).length;
            const totalQuestions = student.answers.length;
            student.score = totalQuestions > 0 ? Math.round((correctAnswers / totalQuestions) * 100) : 0;
        }
    });

    // Средняя оценка
    const avgScore = students_data.reduce((sum, student) => sum + student.score, 0) / students_data.length;
    doc.font(fontPaths.Helvetica)
        .fontSize(12)
        .text(`Средняя оценка по классу — ${avgScore.toFixed(1)}%`);
    doc.moveDown(1);
    currentY = doc.y;

    // Сортируем учеников по номеру
    const sortedStudents = [...students_data].sort((a, b) => parseInt(a.aruco_num) - parseInt(b.aruco_num));

    // Заголовки таблицы
    const headers = ['№', 'Ученики', 'Ответы'];
    if (grading_system.includes('percent')) headers.push('Оценка (%)');
    if (grading_system.includes('five-point')) headers.push('Оценка (5-балльная)');

    // Настройки таблицы
    const colWidths = [1 * 28.35, 5.3 * 28.35, 5 * 28.35];
    if (grading_system.includes('five-point') && grading_system.length === 1) {
        colWidths.push(3.4 * 28.35);
    } else {
        if (grading_system.includes('percent')) colWidths.push(2.6 * 28.35);
        if (grading_system.includes('five-point')) colWidths.push(3.0 * 28.35);
    }

    // Высота строки таблицы
    const rowHeight = 28;
    const headerHeight = rowHeight;

    // Проверяем, поместится ли таблица
    const tableHeight = headerHeight + sortedStudents.length * rowHeight;
    checkPageBreak(tableHeight);

    // Рисуем заголовки таблицы
    doc.font(fontPaths['Helvetica-Bold']).fontSize(10);
    headers.forEach((header, i) => {
        const x = 2 * 28.35 + colWidths.slice(0, i).reduce((a, b) => a + b, 0);
        doc.rect(x, currentY, colWidths[i], headerHeight).fillAndStroke('#f0f0f0', '#000');
        doc.fillColor('black');
        doc.text(header, x + 5, currentY + 3, {
            width: colWidths[i] - 10,
            align: 'center',
            lineBreak: true
        });
    });
    currentY += headerHeight;

    // Рисуем данные таблицы
    doc.font(fontPaths.Helvetica).fontSize(12);
    sortedStudents.forEach((student) => {
        // Проверяем, нужно ли переносить на новую страницу
        if (checkPageBreak(rowHeight)) {
            // Если добавили новую страницу, рисуем заголовки снова
            headers.forEach((header, i) => {
                const x = 2 * 28.35 + colWidths.slice(0, i).reduce((a, b) => a + b, 0);
                doc.rect(x, currentY, colWidths[i], headerHeight).fillAndStroke('#f0f0f0', '#000');
                doc.fillColor('black');
                doc.text(header, x + 5, currentY + 3, {
                    width: colWidths[i] - 10,
                    align: 'center',
                    lineBreak: true
                });
            });
            currentY += headerHeight;
        }

        const answerSymbols = student.answers.map((ans) => (ans ? '+' : '-')).join('');

        const cells = [student.aruco_num, student.name, answerSymbols];
        if (grading_system.includes('percent')) cells.push(student.score.toString());
        if (grading_system.includes('five-point')) cells.push(convertToFivePoint(student.score).toString());

        cells.forEach((cell, i) => {
            doc.rect(2 * 28.35 + colWidths.slice(0, i).reduce((a, b) => a + b, 0), currentY, colWidths[i], rowHeight).stroke();
            doc.fillColor('black');
            doc.text(cell, 2 * 28.35 + colWidths.slice(0, i).reduce((a, b) => a + b, 0) + 5, currentY + 5, {
                width: colWidths[i] - 10,
                align: i === 0 || i === 1 || i === 2 ? 'left' : 'center',
                lineBreak: true
            });
        });

        currentY += rowHeight;
    });

    // Добавляем футер на последнюю страницу
    addFooter();

    doc.end();

    return new Promise((resolve, reject) => {
        const chunks = [];
        doc.on('data', (chunk) => chunks.push(chunk));
        doc.on('end', () => resolve(Buffer.concat(chunks)));
        doc.on('error', (err) => reject(err));
    });
}
module.exports = { generateTestReport };
// Пример входных данных (для теста)
/*
const testData = {
  class_name: "10-А",
  date: "11/04/2025",
  test_name: "Алгебра: Квадратные уравнения",
  grading_system: ['percent', 'five-point'],
  students_data: [
    { aruco_num: "2", name: "Хохлов Тимофей", answers: [false, true, true, false, true] },
    { aruco_num: "1", name: "Иванов Иван", answers: [true, true, false, true, false] },
    { aruco_num: "3", name: "Петров Пётр", answers: [true, true, true, true, true] },
    { aruco_num: "5", name: "Сидорова Анна", answers: [false, true, false, true, false] },
    { aruco_num: "4", name: "Кузнецов Алексей", answers: [true, false, true, false, true] }
  ],
  output_filename: "test_report.pdf"
};
generateTestReport(testData).then(() => console.log("PDF создан!"));
*/
